Leave a tombstone when a domain is deleted

dannyroberts opened the pull request:
https://dimagi-dev.atlassian.net/browse/SAASP-191

##### SUMMARY
This also lets you create a tombstone for any domain name (that doesn't already belong to a domain), like for a domain that was already deleted before this. The presence of a tombstone will prevent anyone from ever creating a new domain with that name. Tombstones are just domains with is_tombstone set to `True`, and anything using `@login_and_domain_required` will 404 when visiting that domain.

dannyroberts committed:
Instead of deleting domain object, set it as a tombstone

and have any request trying to auth against a tombstone 404.

dannyroberts committed:
Add hidden page for managing project tombstones

dannyroberts committed:
Add ability to create a project tombstone through hidden page

orangejenny commented:
This seems similar to introducing soft deletion for domains. Did you consider using the `DocumentType-Deleted` convention instead of introducing a new flag?

It's not clear to me how this will or should interact with various parts of HQ: the domain list dropdown in the top right, the domain selection page you get on login if you don't have a recent domain saved, functions like `Domain.active_for_user`, etc.

Could you explain a bit more what the use case for the new page is? Are we going to create a bunch of tombstones for domains that have been deleted in the past? Is this going to be part of the process when a domain is deleted?

orangejenny commented on corehq/apps/domain/templates/domain/tombstone_management.html:
```diff
@@ -0,0 +1,35 @@
+    <div class="col-xs-12">
+      <h2>{{ current_page.title }}</h2>
+      <p class="help-block">
+        {% blocktrans %}
```
Should this whole file be translated? Or none of it?

orangejenny commented on corehq/apps/domain/views/tombstone.py:
```diff
@@ -0,0 +1,77 @@
+        csv_domain_list = self.cleaned_data.get('csv_domain_list', '')
+        domains = csv_domain_list.split(',')
+
+        if len(domains) > 10:
```
Curious, why does this limitation exist?

dannyroberts replied:
Copy pasted from superuser management, but I guess I don't see any particular reason, so I'll remove it

orangejenny commented on corehq/apps/domain/views/tombstone.py:
```diff
@@ -0,0 +1,77 @@
+
+    def clean(self):
+        csv_domain_list = self.cleaned_data.get('csv_domain_list', '')
+        domains = csv_domain_list.split(',')
```
It's a little bit kinder to regex split on `\s*,\s*` or trim the domains so people can type "domain1, domain2".

dannyroberts replied:
Right, will do

orangejenny commented on corehq/apps/domain/views/tombstone.py:
```diff
@@ -0,0 +1,77 @@
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self.helper = FormHelper()
```
Use `HQFormHelper` instead to get standard proportions, unless there's a reason these are different.

orangejenny commented on corehq/apps/hqadmin/urls.py:
```diff
@@ -53,6 +53,8 @@
     url(r'^auth_as/(?P<username>[^/]*)/(?P<domain>{})/$'.format(new_domain_re),
         AuthenticateAs.as_view(), name=AuthenticateAs.urlname),
     url(r'^superuser_management/$', SuperuserManagement.as_view(), name=SuperuserManagement.urlname),
+    url(r'^tombstone_management/$', TombstoneManagement.as_view(), name=TombstoneManagement.urlname),
```
These should be linked in the admin menu.

orangejenny replied:
Ha, I don't blame you, but this is kind of security by obscurity. I think there's rarely if ever a case for magic links. We should permission things appropriately and/or have specific help text on the page about the risks. Since this is internal, I think adding candid help text is the way to go. In general users aren't good about reading help text, but the support team *is* good about it.

orangejenny reviewed (commented):
Code looks reasonable, but I'd like to understand the premise of this change better.

dannyroberts committed:
Respond to PR feedback

dannyroberts committed:
Do not save domain object on subscription save

This causes a document update conflict if you already have the domain in
hand and later save it. Instead, republish the domain to kafka to get
the pillows to reprocess it, since that was the original motivation.

dannyroberts commented:
Hey @orangejenny, thanks for the thoughtful review. Let me see if I can address the bigger questions with the reasoning behind my own choices, and then we can see from there.

> This seems similar to introducing soft deletion for domains. Did you consider using the DocumentType-Deleted convention instead of introducing a new flag?

Yeah it is pretty similar to this. I considered it, but thought this other way was easier to implement. The challenge is relatively small though now that I think harder about it: it was mostly that I wanted to make sure when creating a domain that there's no existing domain OR domain tombstone (whether that's `is_tombstone: true` or `doc_type: Domain-Deleted`) with the same name, and using `is_tombstone: true` would give the right behavior here for free without having to specifically check for it. Querying for `Domain-Deleted` objects in the db would then also require its own view or a change to an existing view.

> It's not clear to me how this will or should interact with various parts of HQ: the domain list dropdown in the top right, the domain selection page you get on login if you don't have a recent domain saved, functions like Domain.active_for_user, etc.

So I believe as part of deleting the domain, we already remove all memberships to it, which is what drives the domain list dropdown, and the other specific examples you mentioned. There are other places though where we have to address this question. What currently comes to mind is when calculating bulk analytics on domains, and syncing to salesforce and stuff like that. I think for now I want to exclude tombstones from that, and also maybe from ever being saved again, since they're supposed to be a final mark, and if the properties on them are to mean anything they should mean a snapshot of the final state (metadata only of course, all other data is actually and intentionally deleted i.e. at the request of the customer).

> Could you explain a bit more what the use case for the new page is? Are we going to create a bunch of tombstones for domains that have been deleted in the past? Is this going to be part of the process when a domain is deleted?

This PR makes it so that when a domain is deleted its domain object is converted to a domain tombstone object (just by setting `is_tombstone: true`), but like you suggested this page is meant to be used to backfill known domains that have been deleted that need a tombstone. While we try to delete all data associated with a domain and we believe that for the most part we do so successfully, we've also had enough times where we realized this or that piece of data wasn't deleted, that it seems irresponsible to allow anyone who knows the name of the former domain to become its new owner (and the owner of any unintentional residual data) just by creating a "new" domain with that name. That's motivation and the only way I'm currently expecting the page to be used.

dannyroberts committed:
Fix tests by allowing full domain deletion in tests

orangejenny commented:
Hey, thanks for the explanation. My main objection is that I think it's worth a more time-consuming implementation, including adding/changing views, for the sake of having a more understandable code base. We already have concepts of deletion, soft deletion, and archiving, and at times there's confusion about the differences between those. I don't think it's good to introduce another concept/vocabulary to describe what's basically soft deletion.

I'm also not 100% confident that everywhere we reference domains is also gated by domain membership (or will be in the future). The enterprise dashboard reports come to mind. Even if the only places you check tombstone status are the `login_and_domain_required` decorator and analytics, it seems cleaner to me to use `Domain-Deleted` docs and then only query that view in one place (domain creation), no other workflow would ever have to worry about them.

Regarding the UI, it sounds like we might as well just make this part of the command to delete a domain. And then maybe have a dev do a one-time creation of tombstone/deleted domains for domains we've deleted in the past. That said, I don't have a major problem with the UI, the weight of a new admin UI is less than the weight of a new flag/concept for domains.

dannyroberts commented:
OK I'm sold. Thanks for the patient persistence. I'll switch to the more conventional `doc_type: Domain-Deleted`. Our domain db is pretty small and it really should be all that much of a hassle to add a small view for these anyway.

dannyroberts committed:
Add and test view and dbaccessors for dealing with Domain-Deleted

dannyroberts committed:
Use `doc_type: Domain-Deleted` instead of `is_tombstone: true`

dannyroberts committed:
Fix tombstone management

dannyroberts committed:
Better describe what happens during delete_domain mgmt command

orangejenny commented:
@dannyroberts Sweet, thanks for being open.

orangejenny commented on corehq/apps/domain/views/tombstone.py:
```diff
@@ -78,23 +76,20 @@ class TombstoneManagementForm(forms.Form):
         super().__init__(*args, **kwargs)
 
-        self.helper = FormHelper()
+        self.helper = HQFormHelper()
```
Sorry, wasn't clear earlier: this lets you delete the three lines below: https://github.com/dimagi/commcare-hq/blob/868148af1ef1c32dcadfca265866ad4bd53fa1e8/corehq/apps/hqwebapp/crispy.py#L22-L25

dannyroberts replied:
Ah nice thanks

orangejenny commented on corehq/apps/domain/templates/domain/tombstone_management.html:
```diff
@@ -13,22 +13,39 @@ <h2>{{ current_page.title }}</h2>
+              {% endif %}
+            {% else %}
+              Multiple Projects with this name!
+              {% regroup projects by doc_type as doc_type_list %}
```
Interesting, I didn't know about `regroup`.

dannyroberts committed:
Add test util to clear out domains with given names

dannyroberts committed:
Better handle attempts to create previously deleted domains through the UI

dannyroberts committed:
Remove redundant styling from domain tombstone crispy form

dannyroberts committed:
Fix domain name reuse in tests (corehq.apps.users.tests.create)

dannyroberts committed:
Fix domain name reuse in tests (corehq.apps.users.tests.fixture_status)

dannyroberts committed:
Fix domain name reuse in tests (corehq.form_processor.tests.test_basics)

snopoke commented on corehq/apps/domain/models.py:
```diff
@@ -943,9 +954,18 @@ def copies(self):
 
-    def delete(self):
+    def delete(self, leave_tombstone=False):
+        if not leave_tombstone and not settings.UNIT_TESTING:
```
why not make `leave_tombstone` default to `True`? is it just that you'd need to update all the unit tests?

dannyroberts replied:
There was that, and also just that by far the more frequent call in our code base (because of tests) is `leave_tombstone=False`. It's called maybe twice anywhere with `leave_tombstone=True`, and I thought that that would be a nice thing to make explicit in those places for readability. (And then this line is to prevent new code from deleting domains in production _without_ also leaving a tombstone.)

dannyroberts commented:
Tests are passing except for stickler's complaints about the format of couchdb map/reduce js files. (Stickler should really `ignore _design/*/*.js` 🤔 )

orangejenny commented:
@dannyroberts It looks straightforward to add an `.eslintignore`: https://eslint.org/docs/user-guide/configuring#eslintignore

dannyroberts committed:
Add "Manage deleted domains" page to admin interface

dannyroberts committed:
Ignore couchdb map/reduce view files in stickler

orangejenny commented on corehq/apps/domain/_design/views/deleted_domains/map.js:
```diff
@@ -0,0 +1,5 @@
+function (doc) {
```
Hrm, it's still kind of annoying that this message pops up. Another option is to put `/* eslint-disable */` at the top of design doc files, maybe that would be better? (Obviously feel free to call this out of scope.)

dannyroberts commented:
Thanks @orangejenny that didn't end up working but I found what did (in https://stickler-ci.com/docs), and it looks like it worked

dannyroberts commented:
OK! Finally back to ✅. Mind reviewing the last few commits?

dannyroberts committed:
Fix tombstone_management.html to correctly identify Project Tombstones
